### PR TITLE
Make System.Signal async-signal-safe

### PR DESCRIPTION
**Pull request by stepancheg**
It is not safe to call `malloc` or `pthread_mutex_lock` from signal
handler.

CC @mattpolzin

### CODE REVIEW

**Comment by mattpolzin** [comment]
Ah, this makes sense. I did a bit more reading and it sounds like another solution would have been wrapping my existing code in a new thread. Are there disadvantages to that from what you know?

I mostly ask for edification, but a little also because your changes do change the semantics of the code. Before the code gave back signals in the order they occurred and that is no longer the case. Is that really that bad? Probably not. My ring buffer idea might have been over-complicated. Another difference is an arbitrary cap on signal numbers allowed at 32. This handles all the common ones, and as far as I know it handles the ones the respective Idris code exposes, but it also falls short of the NSIG constant currently defined in at least one Linux signals.h I could find (I found one defined as 64).

**Comment by mattpolzin** [comment]
So to summarize the differences introduced here:
1. No deadlocks in the new code (clearly good).
2. The old code was bounded by the number of consecutive signals it collected whereas the new code is bounded by the signum itself. 
3. The new code allocates a bit more space but undoubtedly no big deal in practice. 
4. The new code does not retain information on the order in which signals occur. 

Regarding bullet 2, the old code was intended to be made to expand if needed allowing for unbounded number of consecutive signals once fully fledged. Also on the same bullet, the new code sets a signum cap of 32 which is technically below some systems' NSIG of 64, though maybe that's never an issue in practice as POSIX systems don't tend to define any userland signals above 32. 

Regarding bullet 3, you use longs which allow the new code to collect higher counts of individual signals than the original code. Is that worth it in practice, though? Seems like an unsigned int would be way more than enough. Of course, the original code used a signed int for no good reason.

**Comment by stepancheg** [comment]
> Before the code gave back signals in the order they occurred and that is no longer the case.

>> If multiple standard signals are pending for a process, the order in which the signals are delivered is unspecified.

From [signal(7)](https://man7.org/linux/man-pages/man7/signal.7.html). Attempt to deliver signals in order might give users false sense of signals being delivered in order, which might lead to harder to debug issues.

> Seems like an unsigned int would be way more than enough.

I actually think there's practically no need to have the signal count: it a signal was delivered twice, then it should be collected only once, because:
* practically there's no need, it is not a queue
* in OS implementation there's no signal queue, only signal mask. I. e. if we send two signals very quickly, signal handler will be triggered only once

>> Standard signals do not queue.  If multiple instances of a standard signal are generated while that signal is blocked, then only one instance of the signal is marked as pending


So if this PR is merged, I want to do a follow-up to replace counts with a bitmask.

For this PR I tried to do as few semantic changes as possible.

> The new code allocates a bit more space but undoubtedly no big deal in practice

Allocating 64 * 32 (2K) is really not an issue. But the possible fix is to allocate the array on the first invocation of `collect_signal`.

> I did a bit more reading and it sounds like another solution would have been wrapping my existing code in a new thread. Are there disadvantages to that from what you know?

I didn't get what you meant. Signal is delivered to a random thread. It is possible to force delivery of a signal to a specific thread, but it is very OS-specific, and signal handler is still need to be async-signal-safe.

**Comment by mattpolzin** [comment]
>> I did a bit more reading and it sounds like another solution would have been wrapping my existing code in a new thread. Are there disadvantages to that from what you know?

> I didn't get what you meant. Signal is delivered to a random thread. It is possible to force delivery of a signal to a specific thread, but it is very OS-specific, and signal handler is still need to be async-signal-safe.

What I meant was spin up a thread in the signal handler to do blocking work on that new thread. This seems to be a common way of handling work that might block or take a while and results from a signal being handled. I think your use of atomics feels appropriate, but since your code changes both the way access is handled (mutex vs. atomic) and also the way signals are collected, it occurred to me there might be a smaller footprint solution. I realize now, based on the rest of your latest response that your changes around the semantics of collecting signals were intentional anyway.

> - practically there's no need, it is not a queue
> - in OS implementation there's no signal queue, only signal mask. I. e. if we send two signals very quickly, signal handler will be triggered only once

This is true while a signal is pending at the OS level (i.e. between two cycles where user code is run), but the way this signal handling currently works it introduces a second layer -- the handler might be invoked by the system and in turn collect signals more than once before Idris code handles the collected signals. For that reason, there _is_ a meaning to the order in which signals are collected by us even though there is no meaning to the order in which two pending signals are sent to us from the system. There is no guarantee that an Idris program will handle collected signals every time the signal handler is called by the system.

That said, what if we collect both a SIGTRAP and a SIGCONT between two times when an Idris program choses to handle signals that have been collected? I think it would be important to know which one was collected first, assuming they were sent during separate transitions to user-mode execution.

Reading farther down `singal(7)` docs (first time I've made it this far myself), I see that if we want to expose realtime signal support in the Idris bindings at some point, we will need to support both queuing and signals with numbers 32-64. I am not sure if/when we will support that, but it would be interesting to rip out an ability to queue signals only to need to reintroduce it later :/ 

Thoughts?

**Comment by stepancheg** [comment]
> What I meant was spin up a thread in the signal handler to do blocking work on that new thread.

The simpler approach might be blocking `waitForSignal` call, which could be called from any thread.

> I see that if we want to expose realtime signal support

Realtime signals only work on Linux, and I literally never seen them being used in real applications. Perhaps they don't need to be in Idris core library at all. Especially while Idris standard library lacks basic building blocks.

You initial PR explains the need of signals: handle Ctrl-C from keyboard. Additionally some might want to handle SIGTERM. That covers 99.99% of signal use cases, and 99.9999% percent of all applications. The rest can do it in a separate library.

Also note, Rust standard library (which is quite good in my opinion), does not have signal support at all.

**Comment by mattpolzin** [comment]
> The simpler API might be blocking waitForSignal call, which could be called from any thread.

I believe we are still talking about different things. I am suggesting something in the signal handler, not something related to waiting for a signal.

This is the advice I've read elsewhere:
> If you have code like this inside a signal handler, you either need to rework it to simplify the handler, or have the handler initiate a thread (or signal an existing one) to perform the task.

I'd say your approach amounts to simplifying the handler, but I was talking about the possibility of spinning up a new thread with in the handler.

You've got me convinced to keep the standard library implementation simple and not bother with nearly unused things like realtime signals.

I also think you've convinced me that queuing signals in any capacity or knowing how many times a signal has occurred is unnecessary or even ill-advise. Although I still maintain that the way this library is designed means there _can_ be meaningful queuing of signals such as for debugging, maybe we should remove those signals from the Idris bindings altogether and let that be the concern of another library if needed as you suggest.

**Comment by stepancheg** [comment]
> have the handler initiate a thread

OK, I got it now.

It is not trivial to create a thread from signal handler. `pthread_create` is not a signal-safe function.

You need to have another thread waiting for a signal event. And that can be implemented by having `waitForSignal` building block.

**Comment by mattpolzin** [comment]
I am sold on your changes here, but I think the documentation in the Idris file needs updating to reflect what is actually happening now. The downside to updating it is that if you submit another PR that actually switches from counting signal occurrences to just having a mask of occurrences then the Idris code won't only have outdated documentation it will also be a kind of lame interface to what is actually underneath; by which I mean the final direction you suggest heading seems to warrant rethinking the Idris interface, not just the C code. 

Maybe the Idris file is ok for now if you are definitely sure you'll be submitting a follow up PR and wanting to make changes to the Idris file at that point anyway.

**Comment by stepancheg** [comment]
Updated the doc.

File: libs/base/System/Signal.idr [hunk 1]
```diff
@@ -193,12 +193,8 @@ defaultSignal sig = do
 ||| of your program to retrieve (and mark as handled) the next
 ||| signal that was collected, if any.
 |||
-||| Multiple signals will be collected and can then be retrieved
-||| in the order they were received by multiple calls to
-||| `handleNextCollectedSignal`.
-|||
-||| You can call `handleManyCollectedSignals` to get a List of
-||| pending signals instead of retrieving them one at a time.
+||| Signals are not queued, so the return order or signals is
```
Comment (mattpolzin) [inline libs/base/System/Signal.idr:196]:
```suggestion
||| Signals are not queued, so the return order of signals is
```

**Review by mattpolzin — APPROVED** [review]
Thanks for fixing my attempt at this! I think I hit a deadlock at one point already using my original attempt a few weeks back so removing that problem is much appreciated.

File: libs/base/System/Signal.idr (hunk 1 above)
Comment (stepancheg) [reply libs/base/System/Signal.idr]: Applied. Thanks!

**Comment by mattpolzin** [comment]
@gallais would you be willing to merge this or take a look at it yourself? I feel very confident that the code I originally wrote that is being fixed here results in deadlocks (a matter of when, not if).

Sorry for the ping, I just know you are most active in merging PRs right now.

**Comment by mattpolzin** [comment]
This PR was probably ready to go. I'd love it if you wanted to reopen it @stepancheg.

Based on other PR activity, I gather that Stepan has decided not to continue contributing to the project (which I think is a loss for all because a lot of the PRs just closed made a lot of sense to me even if I don't have the C chops to have come up with them myself), but whether or not that is accurate I'd still like to see the fixes from this PR get merged if possible @gallais .

**Comment by gallais** [comment]
I am always very happy to defer to more knowledgeable people when it comes to
deciding whether the C-related PRs are ready to be merged.

As for the merging of this specific one, I think it would be rude to Stepan
to merge it unless they're OK with it. So I'm crossing my fingers it gets
re-opened in the near future and I can press the magic button.

**Comment by stepancheg** [comment]
I'm back.

**Pull request merged**